File: ABCBS_py/abcbs.py
"""

Description: Python implementation of Anytime bounded conflict-based search (Type 1)

Author: Yue Yang (yuey9923@gmail.com)

"""

import argparse
import yaml
from itertools import combinations
from copy import deepcopy
import time
import random
import logging
from Environment import Environment, Constraints

class HighLevelNode(object):
    def __init__(self):
        self.solution = {}
        self.constraint_dict = {}
        self.cost = 0

    def __lt__(self, other):
        return self.cost < other.cost

class ABCBS(object):
    def __init__(self, environment, logger, limited_time = 30, is_anytime = False, w_h = 1.1, w_l = 1.1, gamma = 0.985):
        self.env = environment
        self.is_anytime = is_anytime
        self.logger = logger
        self.open_set = set()

        # Anytime things
        self.focal_list = []
        self.limited_time = limited_time
        self.gamma = gamma # Decay of bound
        self.w_h = w_h / self.gamma # Bound of high level
        self.w_l = w_l

    def updateFocalBound(self, bound):
        for i in range(len(self.focal_list)):
            node = self.focal_list[i]
            if node.cost > bound:
                self.focal_list.pop(i)

    def updateLowerBound(self, old_bound, new_bound):
        for new_node in self.open_set:
            if (new_node.cost > old_bound) and (new_node.cost < new_bound):
                for i in range(len(self.focal_list)):
                    node = self.focal_list[i]
                    if len(new_node.constraint_dict) <= len(node.constraint_dict):
                        self.focal_list.insert(i, new_node)
                        break
                    if i == len(self.focal_list) - 1:
                        self.focal_list.append(new_node)
                        break
                if len(self.focal_list) == 0:
                    self.focal_list.append(new_node)

    def getNextBound(self):
        return_thing = self.w_h * self.gamma
        if return_thing < 1.001:
            self.logger.info("w_h < 1.001. Have been most optimal")
            return "OPTIMAL"
        return return_thing

    def open_focal_initialization(self, solutions):
        # Construct start node in high level
        start = HighLevelNode()

        # 1. Node.constraint_dict initialization
        start.constraint_dict = {}
        for agent in self.env.agent_dict.keys():
            start.constraint_dict[agent] = Constraints()

        # 2. Node.solution initialization
        start.solution = self.env.compute_solution(self.start_time, self.limited_time)
        # Time exceeded, go outside ~
        if start.solution == "TIME":
            if len(solutions) == 0:
                self.logger.warning("Time exeeded. No solution found.")
                return "OVER", {}
            else:
                return "OVER", solutions[len(solutions) - 1]

        # if start.solution == "OCCUPY":
        #     continue

        if not start.solution:
            return "OVER", {}

        # 3. Node.cost initialization
        start.cost = self.env.compute_solution_cost(start.solution)
        print("start.cost: ", start.cost)

        # OPEN & FOCAL initialization
        self.open_set |= {start}
        self.focal_list.append(start)

        return None, None


    def search(self, N=0, agent_num=0, obstacle_prob=0):
        # Calc consumed time for the first success
        is_first_succ = True
        succ_start_time = time.time()
        # Find how many solutions
        num_sol = 1
        # All found solutions
        solutions = []
        # Time limitation iteration
        self.start_time = time.time()
        # Initialize OPEN & FOCAL
        msg, return_data = self.open_focal_initialization(solutions)
        if msg == "OVER":
            return return_data
        # Keep running if anytime limitation is not over
        while True:
            # If there's no node, initialize again
            if len(self.open_set) == 0 or len(self.focal_list) == 0:
                self.logger.info("One new iteration for a new Solution")
                msg, return_data = self.open_focal_initialization(solutions)

            if msg == "OVER":
                return return_data
            # Tighten the bound
            w_h = self.getNextBound()
            if w_h == "OPTIMAL":
                if len(solutions) == 0:
                    return {}
                else:
                    self.logger.info("Optimal solution found.")
                    return solutions[len(solutions) - 1]
            # Update Focal list
            f_min = min(self.open_set).cost
            self.updateFocalBound(w_h * f_min) # Here, f_min is the f(head(OPEN))
            # Has found one solution?
            has_found_sol = False
            while self.focal_list:
                self.logger.debug("Length of focal list: ", len(self.focal_list))
                f_min = min(self.open_set).cost
                # Get the expanded state from FOCAL
                P = self.focal_list[0]
                # Delete the node from both OPEN & FOCAL
                self.open_set -= {P}
                self.focal_list.pop(0)
                # Update environment
                self.env.constraint_dict = P.constraint_dict
                # Get the first conflict from all agents
                conflict_dict = self.env.get_first_conflict(P.solution)
                self.logger.debug("Conflict dict: " + str(conflict_dict))
                # If there's no conflict for all paths of agents, solution if found!
                if not conflict_dict:
                    self.logger.info(str(num_sol) + " solution(s) are found.")
                    self.logger.info("Cost of success nude: " + str(P.cost))
                    if is_first_succ:
                        is_first_succ = not is_first_succ
                        succ_end_time = time.time()
                        self.logger.info("Success! Consumed time: " + str(succ_end_time - succ_start_time))
                        # If not anytime, return the first solution immediately, else continue compute one new solution
                        if not self.is_anytime:
                            return self.generate_plan(P.solution)
                    num_sol += 1
                    has_found_sol = not has_found_sol
                    solutions.append(self.generate_plan(P.solution))
                    self.open_set = set()
                    self.focal_list = []
                if has_found_sol:
                    break
                # Have conflict, generate constraint dict
                constraint_dict = self.env.create_constraints_from_conflict(conflict_dict)
                # Generate 2 son nodes
                for agent in constraint_dict.keys():
                    self.logger.debug("Solve conflict from: " + str(agent))
                    # 1. Extend constraint dict from father node
                    new_node = deepcopy(P)
                    # 2. Add new constrant to constraint dict according to which agent you choose
                    new_node.constraint_dict[agent].add_constraint(constraint_dict[agent]) 
                    # Update environment constraint
                    self.env.constraint_dict = new_node.constraint_dict # A*就是根据self.env.constraint_dict来得到solution的
                    # 3. Update solution with new constraints added. Now we have the third param (agent), 
                    #       it can make the compute_solution only compute path for this agent
                    new_node.solution = self.env.compute_solution(self.start_time, self.limited_time, agent, deepcopy(P.solution))
                    if new_node.solution == "OCCUPY":
                        continue
                    if new_node.solution == "TIME":
                        if len(solutions) == 0:
                            self.logger.warning("Time exceeded. No solution found.")
                            return {}
                        else:
                            return solutions[len(solutions) - 1]
                    if not new_node.solution:
                        continue
                    # 4. Add cost
                    new_node.cost = self.env.compute_solution_cost(new_node.solution)
                    # Add the son node to open set
                    self.open_set |= {new_node}
                    # If it's conform to the standard(cost <= w * f_min), insert it to the focal list by the conflict number
                    self.logger.debug("New_node.cost: " + str(new_node.cost))
                    self.logger.debug("w_h * f_min: " + str(w_h * f_min))
                    if new_node.cost <= w_h * f_min:
                        for i in range(len(self.focal_list)):
                            node = self.focal_list[i]
                            if len(new_node.constraint_dict) <= len(node.constraint_dict):
                                self.focal_list.insert(i, new_node)
                                break
                            if i == len(self.focal_list) - 1:
                                self.focal_list.append(new_node)
                                break
                        if len(self.focal_list) == 0:
                            self.focal_list.append(new_node)
                # Update the focal list because lower bound of open set is changed, sth may come into the open set~
                if len(self.open_set) != 0:
                    f_min_new = min(self.open_set).cost
                    if (len(self.open_set) != 0) and (self.w_h * f_min < self.w_h * f_min_new):
                        self.updateLowerBound(f_min, f_min_new)
        return {}

    def generate_plan(self, solution):
        plan = {}
        for agent, path in solution.items():
            path_dict_list = [{'t':state.time, 'x':state.location.x, 'y':state.location.y} for state in path]
            plan[agent] = path_dict_list
        return plan

def print_parser(parser, opt):
    message = ''
    message += '\n----------------- Options ---------------\n'
    for k, v in sorted(vars(opt).items()):
        comment = ''
        default = parser.get_default(k)
        if v != default:
            comment = '\t[default: %s]' % str(default)
        message += '{:>25}: {:<30}{}\n'.format(str(k), str(v), comment)
    message += '----------------- End -------------------\n'
    print(message)

def main():
    # Parser initialization
    parser = argparse.ArgumentParser(description="Anytime BCBS")
    parser.add_argument("--input", type=str, default="input.yaml", help="input file containing map and obstacles")
    parser.add_argument("--output", type=str, default="output.yaml", help="output file path")
    parser.add_argument("--wh", type=float, default=1.1, help="omega value of high level")
    parser.add_argument("--wl", type=float, default=1.1, help="omega value of low level")
    parser.add_argument("--gamma", type=float, default=0.985, help="bound decay ratio")
    parser.add_argument("--is-anytime", type=bool, default=False, help="use anytime or not")
    parser.add_argument("--anytime-limitation", type=int, default=300, help="anytime limitation")
    parser.add_argument("--is-logger-file", type=bool, default=False, help="output file path")
    parser.add_argument("--logger-file-path", type=str, default="ABCBS_logger.log", help="log file path")
    parser.add_argument("--logger-level", type=str, default="INFO", \
        choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"], help="logging level")
    args = parser.parse_args()
    
    # Logger initialization
    if args.logger_file_path:
        logging.basicConfig(filename=args.logger_file_path, filemode='w', level=args.logger_level)
    else:
        logging.basicConfig(filemode='w', level=args.logger_level)
    logger = logging.getLogger('ABCBS_logger')
    logger.root.setLevel(args.logger_level)

    logger.info(">>>>>>>>>>>>>>>>>>>>>>> ABCBS START <<<<<<<<<<<<<<<<<<<<<<<<<<")
    print_parser(parser, args)

    # Read information of agents, static obstacles from input file
    logger.info("Read inputs ... ")
    with open(args.input, 'r') as input_file:
        try:
            input_info = yaml.load(input_file, Loader=yaml.FullLoader)
        except yaml.YAMLError as exc:
            logger.error(exc)

    # Start and goal positions of different agents. 
    # e.g. [{'start': [1, 2], 'goal': [1, 2], 'name': 'agent0'}, {'start': [1, 3], 'goal': [1, 5], 'name': 'agent1'}]
    agents = input_info['agents'] 

    # Dimension of the map. 
    # e.g. [32, 32], [64, 64], etc.
    dimension = input_info["map"]["dimensions"]

    # Positions (tuple) of obstacles. 
    # e.g. (0, 3) 
    obstacles = input_info["map"]["obstacles"] 

    # Start to search
    env = Environment(dimension, agents, obstacles, logger, w_l = args.wl)
    cbs = ABCBS(env, logger, args.anytime_limitation, args.is_anytime, args.wh, args.wl, args.gamma)
    logger.info("Start to search ...")
    solution = cbs.search()
    logger.info("Searching end!")
    if not solution:
        logger.warning("Solution NOT found!")
        logger.info(">>>>>>>>>>>>>>>>>>>>>>> ABCBS END <<<<<<<<<<<<<<<<<<<<<<<<<<")
        return
    else:
        logger.info("Solution found!")

    output = {}
    output['schedule'] = solution

    # Write solution to output file
    logger.info("Write solution to output file ... ")
    with open(args.output, 'w') as output_yaml:
        yaml.safe_dump(output, output_yaml)

    logger.info(">>>>>>>>>>>>>>>>>>>>>>> ABCBS END <<<<<<<<<<<<<<<<<<<<<<<<<<")

if __name__ == "__main__":
    main()

File: ABCBS_cpp/ABCBS.cpp
//
// Created by Yue Yang on 2020/11/24.
//

#include "ABCBS.h"

ABCBS::ABCBS(const Environment &env, bool isAnytime, int limitedTime, double gamma, double wH, double wL) : env(env),
                                                                                                            is_anytime(
                                                                                                                    isAnytime),
                                                                                                            limited_time(
                                                                                                                    limitedTime),
                                                                                                            gamma(gamma),
                                                                                                            w_h(wH),
                                                                                                            w_l(wL) {
    this->w_h = w_h / gamma;
}

void ABCBS::updateFocalBound(double bound) {
    for (int i = 0; i < this->focal_list.size(); ++i) {
        HighLevelNode node = HighLevelNode();
        try {
                node = this->focal_list.at(i);
            }
            catch (std::out_of_range &exc) {
                std::cerr << exc.what() << " Line:" << __LINE__ << " File:" << __FILE__  << std::endl;
            }

        if (node.getCost() > bound) {
            this->focal_list.erase(this->focal_list.begin() + i);
        }
    }
}

void ABCBS::updateLowerBound(double old_bound, double new_bound) {
    for (auto const& new_node : this->open_set) {
        if ((new_node.getCost() > old_bound) and (new_node.getCost() < new_bound)) {
            for (int i = 0; i < this->focal_list.size(); ++i) {
                HighLevelNode node = HighLevelNode();
                try {
                        node = this->focal_list.at(i);
                    }
                    catch (std::out_of_range &exc) {
                        std::cerr << exc.what() << " Line:" << __LINE__ << " File:" << __FILE__  << std::endl;
                    }

                if (new_node.getConstraintDict().size() <= node.getConstraintDict().size()) {
                    this->focal_list.insert(this->focal_list.begin() + i, new_node);
                    break;
                }
                if (i == (this->focal_list.size() - 1)) {
                    this->focal_list.push_back(new_node);
                    break;
                }
            }

            if (this->focal_list.size() == 0) {
                this->focal_list.push_back(new_node);
            }
        }
    }
}

std::tuple<bool, double> ABCBS::getNextBound() {
    double return_thing = this->w_h * this->gamma;
    if (return_thing < 1.001) {
        return std::make_tuple(true, 0);
    }
    return std::make_tuple(false, return_thing);
}

std::tuple<bool, std::map<std::string, std::vector<std::map<std::string, int>>>> ABCBS::open_focal_initialization(
        std::vector<std::map<std::string, std::vector<std::map<std::string, int>>>>& solutions) {
//    Construct start node in high level
    HighLevelNode start = HighLevelNode();

//    1. Node.constraint_dict initialization
    std::map<std::string, Constraints> constraint_dict;
    for (auto const& agent_info : this->env.getAgentDict()) {
        std::string agent = agent_info.first;
        try {
                constraint_dict[agent] = Constraints();
            }
            catch (std::out_of_range &exc) {
                std::cerr << exc.what() << " Line:" << __LINE__ << " File:" << __FILE__  << std::endl;
            }

        start.setConstraintDict(constraint_dict);
    }

//    2. Node.solution initialization
    ComputeSolution computeSolution = ComputeSolution(this->env);
    std::tuple<bool, bool, bool, std::map<std::string, std::vector<State>>>  outcome = \
    computeSolution.compute_solution(this->start_time, this->limited_time);
    start.setSolution(std::get<3>(outcome));

//    Time exceeded, go outside ~
    if (std::get<1>(outcome)) {
        if (solutions.size() == 0) {
            std::cout << "Time exeeded. No solution found." << std::endl;
            std::map<std::string, std::vector<std::map<std::string, int>>> tmp;
            return std::make_tuple(true, tmp);
        } else {
            try {
                    return std::make_tuple(true, solutions.at(solutions.size() - 1));
                }
                catch (std::out_of_range &exc) {
                    std::cerr << exc.what() << " Line:" << __LINE__ << " File:" << __FILE__  << std::endl;
                }
        }
    }

    if (std::get<0>(outcome)) {
        std::map<std::string, std::vector<std::map<std::string, int>>> tmp;
        return std::make_tuple(true, tmp);
    }



//    3. Node.cost initialization
    start.setCost(this->env.compute_solution_cost(start.getSolution()));

//    OPEN & FOCAL initialization
    this->open_set.insert(start);
    this->focal_list.push_back(start);
    std::map<std::string, std::vector<std::map<std::string, int>>> tmp;

    return std::make_tuple(false, tmp);
}



std::map<std::string, std::vector<std::map<std::string, int>>> ABCBS::generate_plan(
        const std::map<std::string, std::vector<State>> &solution) {
    std::map<std::string, std::vector<std::map<std::string, int>>> plan;
    for (auto const& ag_pth : solution) {
        std::string agent = ag_pth.first;
        std::vector<State> path = ag_pth.second;
        std::vector<std::map<std::string, int>> path_dict_list;
        for (auto const& state : path) {
            std::map<std::string, int> pt = {
                    {"t", state.getTime()},
                    {"x", state.getLocation().getX()},
                    {"y", state.getLocation().getY()}
            };
            path_dict_list.push_back(pt);
        }
        try {
                plan[agent] = path_dict_list;
            }
            catch (std::out_of_range &exc) {
                std::cerr << exc.what() << " Line:" << __LINE__ << " File:" << __FILE__  << std::endl;
            }

    }
    return plan;
}




std::tuple<bool, std::map<std::string, std::vector<std::map<std::string, int>>>> ABCBS::search() {
//    Calc consumed time for the first success
    bool is_first_succ = true;
    int succ_start_time = time(NULL);

//    Find how many solutions
    int num_sol = 1;

//    All found solutions
    std::vector<std::map<std::string, std::vector<std::map<std::string, int>>>> solutions;

//    Time limitation iteration
    setStartTime(time(NULL));
//    int start_time = time(NULL);

//    Initialize OPEN & FOCAL
    std::tuple<bool, std::map<std::string, std::vector<std::map<std::string, int>>>>  msg_sol = \
    this->open_focal_initialization(solutions);

    if (std::get<0>(msg_sol)) {
        return std::make_tuple(false, std::get<1>(msg_sol));
    }

//    Keep running if anytime limitation is not over
    while (true) {

//        If there's no node, initialize again
        if ((this->open_set.size() == 0) and (this->focal_list.size() == 0)) {
            std::cout << "One new iteration for a new Solution" << std::endl;
            std::tuple<bool, std::map<std::string, std::vector<std::map<std::string, int>>>> msg_sol = \
        this->open_focal_initialization(solutions);

        }


        if (std::get<0>(msg_sol)) {
            return std::make_tuple(false, std::get<1>(msg_sol));
        }
//        Tighten the bound
        std::tuple<bool, double> w_h_info = this->getNextBound();

        if (std::get<0>(w_h_info)) {
            if (solutions.size() == 0) {
                std::map<std::string, std::vector<std::map<std::string, int>>> tmp;
                return std::make_tuple(true, tmp);
            } else {
                std::cout << "Optimal solution found."  << std::endl;
                try {
                        return std::make_tuple(false, solutions.at(solutions.size() - 1));
                    }
                    catch (std::out_of_range &exc) {
                        std::cerr << exc.what() << " Line:" << __LINE__ << " File:" << __FILE__  << std::endl;
                    }

            }
        }


//        Update Focal list
        double f_min = std::numeric_limits<int>::max();
        for (auto const& node : this->open_set) {
            if (node.getCost() < f_min) {
                f_min = node.getCost();
            }
        }


        this->updateFocalBound(w_h * f_min); // Here, f_min is the f(head(OPEN))

//        Has found one solution?
        bool has_found_sol = false;

        while (this->focal_list.size() != 0) {
            std::cout << "Length of focal list: " << this->focal_list.size() << std::endl;

            double f_min = std::numeric_limits<int>::max();

            for (auto const& node : this->open_set) {
                if (node.getCost() < f_min) {
                    f_min = node.getCost();
                }
            }

//            Order Focal list if not efficiently reusable (Here, it is. So do nothing)

//            Get the expanded state from FOCAL
            HighLevelNode P = HighLevelNode();
            try {
                    P = this->focal_list.at(0);
                }
                catch (std::out_of_range &exc) {
                    std::cerr << exc.what() << " Line:" << __LINE__ << " File:" << __FILE__  << std::endl;
                }

//            Delete the node from both OPEN & FOCAL
            this->open_set.erase(P);
            this->focal_list.erase(this->focal_list.begin());

//            Update environment
            this->env.setConstraintDict(P.getConstraintDict());

//            Get the first conflict from all agents
            std::tuple<bool, Conflict> first_conf = this->env.get_first_conflict(P.getSolution());

//            No conflict found
            if (!std::get<0>(first_conf)) {
                std::cout << ">>> " << num_sol << " solution(s) are found." << std::endl;
                std::cout << ">>> Cost of success node: " << P.getCost() << std::endl;

                if (is_first_succ) {
                    is_first_succ = !is_first_succ;
                    int succ_end_time = time(NULL);

                    std::cout << ">>> Success! Consumed time: " << succ_end_time - succ_start_time  << std::endl;

//                    If not anytime, return the first solution immediately, else continue compute one new solution
                    if (! this->is_anytime) {
                        return std::make_tuple(false, this->generate_plan(P.getSolution()));
                    }
                }

                num_sol += 1;
                has_found_sol = !has_found_sol;
                solutions.push_back(this->generate_plan(P.getSolution()));
                this->open_set.clear();
                this->focal_list.clear();
            }

//            Conflict found
            if (has_found_sol) {
                break;
            }

//            Have conflict, generate constraint dict
            std::map<std::string, Constraints> constraint_dict = \
            this->env.create_constraints_from_conflict(std::get<1>(first_conf));

//            Generate 2 son nodes
            for (auto const& agent_info : constraint_dict) {
                std::string agent = agent_info.first;

//                1. Extend constraint dict from father node
                HighLevelNode new_node = HighLevelNode(P);


//                2. Add new constrant to constraint dict according to which agent you choose
                try {
                        std::map<std::string, Constraints> constraintDict = new_node.getConstraintDict();
                        Constraints constraints = constraintDict.at(agent);
                        constraints.add_constraint(constraint_dict.at(agent));
                        constraintDict[agent] = constraints;
                        new_node.setConstraintDict(constraintDict);
                    }
                    catch (std::out_of_range &exc) {
                        std::cerr << exc.what() << " Line:" << __LINE__ << " File:" << __FILE__  << std::endl;
                    }


//                Update environment constraint
                this->env.setConstraintDict(new_node.getConstraintDict()); // A*就是根据self.env.constraint_dict来得到solution的


                /*
                 * 3. Update solution with new constraints added. Now we have the third param (agent),
                 * it can make the compute_solution only compute path for this agent
                 * */
                ComputeSolution computeSolution = ComputeSolution(this->env);
                std::tuple<bool, bool, bool, std::map<std::string, std::vector<State>>> compute_result = \
                computeSolution.compute_solution(this->start_time, this->limited_time, agent, &P.getSolution());

                if (std::get<2>(compute_result)) {
                    continue;
                }
                if (std::get<1>(compute_result)) {
                    if (solutions.size() == 0) {
                        std::cout << "Time exceeded. No solution found."  << std::endl;
                        std::map<std::string, std::vector<std::map<std::string, int>>> tmp;
                        return std::make_tuple(true, tmp);
                    } else {
                        return std::make_tuple(false, solutions.at(solutions.size() - 1));
                    }
                }

                if (std::get<0>(compute_result)) {
                    continue;
                }


                new_node.setSolution(std::get<3>(compute_result));

//                4. Add cost
                new_node.setCost(this->env.compute_solution_cost(new_node.getSolution()));

//                Add the son node to open set
                this->open_set.insert(new_node);

//                If it's conform to the standard(cost <= w * f_min), insert it to the focal list by the conflict number
                if (new_node.getCost() <= w_h * f_min) {
                    for (int i = 0; i < this->focal_list.size(); ++i) {
                        HighLevelNode node = this->focal_list.at(i);
                        if (new_node.getConstraintDict().size() <= node.getConstraintDict().size()) {
                            this->focal_list.insert(this->focal_list.begin() + i, new_node);
                            break;
                        }
                    }
                    if (this->focal_list.size() == 0) {
                        this->focal_list.push_back(new_node);
                    }
                }


//                Update the focal list because lower bound of open set is changed, sth may come into the open set~
                if (this->open_set.size() != 0) {
                    double f_min_new = std::numeric_limits<int>::max();
                    for (auto const& node : this->open_set) {
                        if (node.getCost() < f_min_new) {
                            f_min_new = node.getCost();
                        }
                    }

                    if ((this->open_set.size() != 0) and (this->w_h * f_min < this->w_h * f_min_new)) {
                        this->updateLowerBound(f_min, f_min_new);
                    }

                }
            }

        }


    }

    std::map<std::string, std::vector<std::map<std::string, int>>> tmp;
    return std::make_tuple(true, tmp);
}

void ABCBS::setStartTime(int startTime) {
    start_time = startTime;
}

